Feature/config

kevinxufs opened the pull request.

Commit:
repro-catalogue tool auto commit at 20200630-103512

Commit:
repro-catalogue tool auto commit at 20200630-103837

Commit:
repro-catalogue tool auto commit at 20200630-104653

Commit:
repro-catalogue tool auto commit at 20200630-112922

Commit:
repro-catalogue tool auto commit at 20200630-115238

Commit:
repro-catalogue tool auto commit at 20200630-121105

Commit:
repro-catalogue tool auto commit at 20200630-121129

Commit:
modified some main functions

Commit:
refined config tests

kevinxufs commented:
What this PR does:

- Allows the user to create config files
- Updates the parser to look for config files if they exist
- created tests for the new config functionality.

Current problems:
1) the config tests currently work, but I got a failed test in engage_test for some reason.

![image](https://user-images.githubusercontent.com/48526846/88162397-3558b780-cc09-11ea-86ef-28480004db29.png)

Note that this doesn't seem to be a problem with my VM, the same test still runs on the master branch.


2) Will want to refactor some of the code in the tests.

Commit:
cleaned the code a bit

edaub commented on catalogue/parser.py:
```diff
@@ -1,125 +1,145 @@
+        description="",
+        formatter_class=argparse.RawTextHelpFormatter)
+
+    main_dict = {'input_data' : r'/home/kevinxu/TuringDataStories/cata_test/input_data',
```
You need to convert these back to relative paths so that they work on anyone's computer.

edaub commented on .idea/misc.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
You should delete these xml files from the repo.

edaub commented on catalogue/compare.py:
```diff
@@ -11,6 +11,9 @@ def compare(args):
     Prints results on the command line.
     """
 
+    if type(args.csv) == float:
```
Usually the preferred way of doing this is:

```
if isinstance(args.csv, float):
```

However, I recall discussing that the "correct" behavior should be for args.csv to be converted to `None` if no value is provided. So in that case, `if args.csv is None:` would be correct.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,41 @@
+from argparse import Namespace
+import pandas as pd
+
+# To test, create an example namespace, e.g:
```
You should delete this from the codebase.

edaub commented on catalogue/engage.py:
```diff
@@ -1,162 +1,172 @@
+import os
+import json
+import git
+import numpy
```
I don't think you use numpy or the math module here? Or did I miss something?

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,41 @@
+def config(args):
+
+    #location of config file
+    config_loc = 'catalogue_config.csv'
```
This is defined twice in two separate places, which is bad practice. Given that it needs to be pre-defined in order to work, I think it is best to make it a variable available in the utils module and then imported into the config and parser files.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,41 @@
+        print('Creating config file')
+
+    else:
+        dict = pd.read_csv(config_loc, header=None, index_col=0, squeeze=True).to_dict()
```
This code is used twice, so it should be a separate function (probably in utils?). While it is one line, we discussed the problem of dealing with the `csv = None` case should be part of the reader function (see comment in `compare.py` so that it works as desired, meaning that it should definitely be a function.

edaub replied:
Also, I'm confused why you are checking for a config file again here -- is it just to print out the old values if they exist? This seems unnecessary given all the work you did to prioritize the different sources of values in the parser.

edaub replied:
And finally, what if the csv isn't formatted correctly? What if there are 3 entries in one row? What if one key value is repeated? Will this still work correctly? Or will you need to raise some kind of error? You should think about how you might handle some cases here, and put them into the unit testing to confirm they fail in the way you expect.

edaub commented on catalogue/parser.py:
```diff
@@ -1,125 +1,145 @@
+
+    config_file_loc = 'catalogue_config.csv'
+
+    if os.path.isfile(config_file_loc):
```
This should be a function, as discussed in the config.py file.

edaub commented on catalogue_config.csv:
```diff
@@ -0,0 +1,5 @@
+input_data,hi
```
I'm confused. Is this here to run the unit test? If so, I would move this to the fixtures directory and integrate it in that way.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,41 @@
+
+    #Check if a config file exists already. If a config file already exists, convert to dictionary and print values
+
+    if os.path.isfile(config_loc) is False:
```
`if not os.path.isfile(config_loc):`

edaub commented on cata_test/input_data/example_data.csv:
```diff
@@ -0,0 +1,2 @@
+hello,world
```
I think you should delete the `cata_test` directory, unless it is used in the testing (and if so, it belongs in the tests directory).

edaub reviewed (changes requested):
Mostly looks good -- some comments inline that I think you need to clean up. Two big things:

1. We discussed how you should make the code to read the config file a separate function, so that you can handle things like `csv = None` once and then not have to worry about fixing it elsewhere. This is one thing I think you should address, in part because it means you can change things that are specific to reading the file without modifying the other code.

2. What happens if your file isn't formatted as you expect? You hand off reading the file and converting to a dictionary to pandas (something I think we need to change once we get everything else working), so (a) how does pandas handle these cases? (b) how should we handle these cases and are they different? and (c) how can we test that if we give it bad inputs, it fails in the desired way?

Anyways, some things to think about as we continue to work on this.

kevinxufs commented:
Thanks a lot @edaub , I'll take a look at some of these changes tomorrow morning before we meet.

Commit:
various deletes, cleaning repo

Commit:
cleaned code, fixing relative paths, adding a function in utils

kevinxufs commented:
So following our chat just now @edaub we agreed I'll focus on the following few things:

1) Look into yml config files and refactor the script to process yml files instead.

2) Establish a process for validating that a given configuration file is correct, when used by either `parser.py` or `config.py`

3) Write a fourth test confirming that the process of reading / checking config files is correct.

4) (Minor fixes with some of the tests, e.g test 2 with creating a fixture config file)

Commit:
changed csv to yaml

Commit:
fixed variable in read_config_file func

kevinxufs commented:
After the latest commit I (hope) that the `catalogue_config` is working once again this time using `yaml files`. It seems fine but I still need to write tests to confirm.

I found it super easy to use yaml files and they were much less of a stress than csv files. In particular, there is no issue with copying over `None` to yaml config files.

I've also cleaned up my functions a bit, adding a new utils function to serve as my dictionary printer as I was doing this printing twice in config.

Commit:
fixed existing_config test and new_config test

kevinxufs commented:
I've fixed two tests in `test_config`

The first test is with generating a new config file. I introduced a new config example and then wrote it in the temporary directory. This allows me to avoid running the config function twice.

The second was with generating a new config file. Since I'm now using a yaml file I had to make the appropriate modifications.

I think there are three more things I need to (would appreciate any steer here @edaub )

1) Write tests for the two new util functions I have added (`read_config_file`, `print_dictionary`)
2) Write some kind of config file validation function and modify my scripts to use config file if and only if the file exists and is validated.
- I need to consider what validated means. I'm thinking of something 'close enough' to the sort of config file that I create
3) Write tests for this config file validation function

Current plan is to get some kind of working version of these 3 by Wednesday afternoon.


Actually there is also a fourth - one of the engage tests is not working (see above) and I'm not sure why.

Commit:
added two tests to test_utils.py

Commit:
added validator function

Commit:
added dictionary check

kevinxufs commented:
The two tests were relatively ok.

The validator function was a bit complicated.

Currently I have thought of five kinds of validations.

I'm expecting a config file (`catalogue_config.yaml) to look something like:

```
catalogue_result: catalogue_results
code: code
csv: null
input_data: input_data
output_data: output_data
```

The five checks:
1) that the result of reading `catalogue_config.yaml`) indeed gives me back a dictionary
2) that the provided keys are a subset of the five main keys 
3) that if the csv file flag is run (i.e. the value of csv is not `None`) then a `.csv` file extension is provided
4) that the config file key values are all strings (instead of e.g. lists)
5) that there are no duplicate keys.

I've currently implemented the first four, but struggling to do the fifth.

That is, I'm not sure how to check whether a yaml file has duplicate keys upon reading. The issue is that when I run something like `yaml.safe_load` on a yaml file with duplicate keys, it just takes the second one without commenting that there were duplicates.

Commit:
added config validation tests

kevinxufs commented:
I've moved things a bit, putting the config validator within `config.py`.

I've now added some tests for the config validator function.

One issue I had was that I wanted to add some yaml files as additional fixtures. Adding them to the fixtures directory (via `conftest`) however messes up a lot of the `catalogue.py` tests. Instead I have created them as seperate yaml files currently in test.

I have 3 yaml files that I have added:
`good_config.yaml`: a yaml file in the format we expect
`bad_config1.yaml`: a config file that gets read as a dictionary (passes condition 1), but is otherwise messed up (fails condition 2-5)
`bad_config2.yaml`: a config file that fails to get read as a dictionary (fails condition 1)

There are still two issues that are worth discussing:

1) Validator fails to check for double key (i.e. i want to be able to take in a yaml file and tell the user if any keys exist twice)
2) engage test error (see above)

kevinxufs commented:
Actually another fix I need to make is to establish why Travis keeps breaking.

I think it is because I am using additional modules (yaml) so I should add that somewhere.

I should also remove pandas.

Commit:
removed pandas requirement, added pyyaml req

kevinxufs committed:
Delete fixture5.yaml

Removed because of test errors

kevinxufs commented:
So I fixed some of the issues with the branch. Now the only problem is with the engage test. Once that is fixed, Travis build should work as well.

Commit:
added fixtures in fixtures directory

Commit:
Merge branch 'feature/config' of https://github.com/alan-turing-institute/repro-catalogue into feature/config

Commit:
moved config file fixtures to fixtures folder

kevinxufs commented:
As a general status update:

Added the test yaml files into the fixtures directory, had to fix a number of tests as a result.

@edaub will be looking into generalising the fixture tests and also taking a look at the one faulty engage test above.

I'll be on leave next week so hopefully we can get this merged the week after I'm back!

edaub committed:
normalize all line endings

Commit:
fixed the engage test by using tmp directories

edaub committed:
Merge remote-tracking branch 'origin/feature/config' into feature/config

edaub committed:
forgot to delete some text in the merge conflict reslution

edaub commented:
Great, I just pushed my commits to normalize the line ending encoding (`config.py` had Windows encodings while the rest of the code base did not). I will look over once more next week and then we should be able to merge. I'm happy to have the config documentation be a separate PR so that we can get this long-running branch merged in!

kevinxufs commented:
Thanks a lot Eric. I'll take a final look monday to see if there's anything I should be clearing up and then I'll work on the documentation as a seperate PR.

radka-j commented on tests/test_config.py:
```diff
@@ -0,0 +1,105 @@
+def test_no_config(test_args, capsys, tmpdir):
+
+    #Ensure no file
+    config_file = os.path.join(tmpdir,'catalogue_config.yaml')
```
very **minor** comment - you don't do anything with `config_file` in this test so I'd remove line 20.

Commit:
added code specific documentation

Commit:
removed redundant config location check

kevinxufs commented:
I have added a bit more indepth documentation for the specific functions (`config validator` and `config`).

There's still some work to add some documentation and usage guidance to the read the docs.

We agreed to add this in a seperate pull request.

We're also still missing parser tests. This was fine previously but as most of the work involved in config sits within parser, we should consider adding this (probably also as a seperate PR)

edaub commented on catalogue/compare.py:
```diff
@@ -11,6 +11,7 @@ def compare(args):
     Prints results on the command line.
     """
 
+
```
Delete this extra empty line.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+            filename = config_dict['csv']
+
+            #change to use os path
+            extension = filename[-4:]
```
I think you forgot to change this to use the `os` module to more robustly extract the extension.

kevinxufs replied:
I have removed this and added this check to catalogue engage where we check for `csv` or `json`

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+
+            #change to use os path
+            extension = filename[-4:]
+            if extension != '.csv':
```
Am I allowed to use a file name with no extension? The CSV functions don't check this, and as long as it determines that the file handle is valid it doesn't complain. My worry is that if I use the tool without using a config file and provide a csv path with no extension, then use a config file later with the existing CSV file it will complain even though the file is perfectly valid. So we probably should decide how we want this to work, and if we do want to enforce an extension we should do that in the CSV functions as it will be more maintainable that way.

kevinxufs replied:
I have added a bit.

If you use a file name with no extension (e.g. 'catalogue_results') as your 'csv' argument, it defaults back to json.

It only creates a csv file if you put an argument for `csv` that also has the correct extension

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+        # check that all config file keys only have string values (i.e. no nested)
+        values_list = list(config_dict.values())
+        for value in values_list:
+            if not isinstance(value, str) and value is not None:
```
I think (don't quote me on this as it is from memory) you can provide a tuple of types to `isinstance` and it will return True if it is one of them. Probably would clean this up a bit and make the logic more clear.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+                valid = False
+                print('Config error: config files are not all strings')
+
+        # check that there are no double keys
```
What does this comment refer to? I think the double key situation is handled however the yaml library does it, so isn't relevant here.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+
+    else:
+        print('Config file identified, checking validity')
+        config_validator(CONFIG_LOC)
```
I don't think you need this here.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+
+    if not os.path.isfile(CONFIG_LOC):
+        print('No previous config file found')
+        print('Creating config file')
```
I think these are redundant since you say later you create a new config file no matter what the outcome is of this check. I think it is fine to just say it once when you create the new one.

kevinxufs replied:
I've removed the second print. I think we should keep the first print since it allows the user to account for the possibility that a config file was found but was not valid.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+            dictionary_printer(dict)
+        else:
+            print('Identified config file is invalid')
+            print('Creating config file')
```
See above.

edaub commented on catalogue/engage.py:
```diff
@@ -98,8 +98,10 @@ def engage(args):
     ---------
     None
     """
+
```
Remove the extra empty lines in this file.

edaub commented on tests/fixtures/bad_config1.yaml:
```diff
@@ -0,0 +1,8 @@
+catalogue_results: catalogue_results
+code: code
+csv: badconfigs.xlsx
```
See above comment on whether we want to require csv extensions.

edaub commented on tests/fixtures/good_config.yaml:
```diff
@@ -0,0 +1,5 @@
+catalogue_results: catalogue_results
+code: code
+csv: null
```
Does this automatically get converted into `None`?

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+    setattr(test_args, "output_data", output_dir)
+    os.chdir(tmpdir)
+
+    #TODO: manually create something (e.g. a fixture) instead of just running it twice
```
Not sure you need a fixture here if this is the only test that uses this object. If you were to write a new test that does use this, then I would make a fixture.

kevinxufs replied:
Do you mean the `test_args` fixture? I think this one was created beforehand.

edaub commented on catalogue/utils.py:
```diff
@@ -21,9 +23,11 @@ def check_paths_exists(args):
     ---------
     Boolean indicating if all filepaths exist.
     """
+    print(args)
```
Do you still want these, or are they for debugging?

edaub commented on catalogue/utils.py:
```diff
@@ -21,9 +23,11 @@ def check_paths_exists(args):
     paths = [value for key, value in vars(args).items()
             if key not in ["command", "func", "csv", "catalogue_results"]]
     path_checks = [os.path.exists(path) for path in paths]
+    print(path_checks)
```
Is this necessary?

edaub commented on catalogue/utils.py:
```diff
@@ -43,3 +47,15 @@ def prune_files(files, dir):
+    return config_data
+
+
+def dictionary_printer(dict):
```
`dict` is a python keyword, I would use a different variable name here.

edaub commented on catalogue/utils.py:
```diff
@@ -43,3 +47,15 @@ def prune_files(files, dir):
+
+def dictionary_printer(dict):
+    for key, value in dict.items():
+        print('{}:{}'.format(key, value))
```
I can be kind of a stickler for typesetting detail, so I would suggest adding a space after the colon for clarity.

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+def test_config_validator(tmpdir, capsys, good_config, bad_config1, bad_config2):
+
+    # validates good config file
+    valid_file = config_validator(good_config)
```
For all of these, you can make your code a bit more compact if you write `assert config_validator(good_config)` or `assert not config_validator(bad_config)` if you like.

kevinxufs replied:
Thanks - I didn't realise that running the assert command would also cause all of the print text of config validator to still be printed. This is of course needed for the later `capsys` checks.

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+    not_dictionary = config_validator(bad_config2)
+    captured = capsys.readouterr()
+    assert 'Config error: yaml file cannot be read as a dictionary' in captured.out
+    assert not not_dictionary
```
You can merge this with the call to `config_validator` above.

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+    assert 'Config error: config files are not all strings'in captured.out
+
+
+    assert not invalid_file
```
Same as above.

edaub commented on tests/test_engage.py:
```diff
@@ -9,6 +9,7 @@
 from catalogue.engage import engage, disengage, git_query
 
 
+
```
Remove extra blank lines.

edaub commented on tests/test_utils.py:
```diff
@@ -24,3 +25,20 @@ def test_check_paths_exists(test_args):
+    dict = read_config_file(good_config)
+    assert dict['code'] == 'code'
+    assert dict['input_data'] == 'input_data'
+    assert dict['csv'] == None
```
I think the preferred way to check if something is NoneType in python is `assert dict['csv'] is None`

edaub commented on tests/test_utils.py:
```diff
@@ -24,3 +25,20 @@ def test_check_paths_exists(test_args):
+    dictionary_printer(dict)
+    captured = capsys.readouterr()
+
+    assert 'a:hi\nb:bye' in captured.out
```
If you change the spacing on the printer, you will need to fix this test.

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+import os.path
+import csv
+import argparse
```
Do you use this module, or just `namespace` (which you import in the next line)?

edaub commented on catalogue/config.py:
```diff
@@ -0,0 +1,122 @@
+import os.path
+import csv
```
I don't think you use this module anymore.

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+import pytest
+
+import csv
+import argparse
```
Do you use argparse, or just namespace?

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+
+import os
+import sys
```
Do you use the sys module?

edaub commented on tests/test_config.py:
```diff
@@ -0,0 +1,102 @@
+import glob
+import pytest
+
+import csv
```
Do you still use the csv module?

edaub commented on tests/test_utils.py:
```diff
@@ -1,13 +1,14 @@
 
 import pytest
+import os
```
Do you use the os module in this file?

edaub commented:
I have noted a number of things in the review comments. Most of them are tiny stylistic things that you can do what you wish with (I only include them here to point out things you may want to be aware of as you get more familiar with writing software).

The one thing I think needs addressing is the check of the csv extension. As I note, if I were to call catalogue without using a config file, it would be perfectly happy saving in a file without a csv extension. Thus, if I call it again using a config file and that perfectly valid csv file as the `csv` argument, it would complain. My feeling is that if we want to check the csv extension, we should _not_ do it here, as this is not required for the `config` function to operate correctly. Rather, checks like that should only be done where they are relevant for that function working correctly to ensure that the code is maintainable (i.e. if I want to change how it saves to csv files, I do not want to have to change anything anywhere else in the code other than the functions that read and write csv files).

Commit:
fixed basic formatting issues

Commit:
dealing with conflicts

Commit:
fixing merge conflicts

kevinxufs commented:
@edaub 
I've taken a look at this comment about csvs and I agree.

My suggestion is to modify the function where the file is saved as a csv. This occurs within engage.

I suggest something like:

```
if the extension of args.csv is a csv file, then save it as a csv file. 

Else, save it as a hash.
```

Commit:
modified csv file checks and added 2 tests

radka-j commented on catalogue/engage.py:
```diff
@@ -1,162 +1,162 @@
+    ---------
+    None
+    """
+    print(args)
```
Why is there a print statement here? I'm guessing it's a left-over from testing?

kevinxufs replied:
Whoops, I've just fixed this now.

Commit:
removed left-over statements from testing

radka-j commented on tests/test_utils.py:
```diff
@@ -24,3 +22,16 @@ def test_check_paths_exists(test_args):
+    dict = {'a': 'hi', 'b' :'bye'}
+    dictionary_printer(dict)
+    captured = capsys.readouterr()
+    assert 'a: hi\nb: bye' in captured.out
```
I get an error running this test - the test works if I remove the spaces from the expected output so that I have `'a:hi\nb:bye'`.
@kevinxufs does this test work for you as is?

kevinxufs replied:
Hi radka. This test seems to work for me. This test was recently modified as Eric suggested that I reformat the `dictionary_printer` function in `utils.py`. Could you check your `utils.py` to see if there is a space in between the curly brackets?

radka-j commented on tests/test_engage.py:
```diff
@@ -142,3 +140,38 @@ def test_engage(git_repo, test_args, capsys):
+
+    print(output_filejson)
+    assert len(output_filecsv) == 0
+    assert len(output_filejson) == 1
```
This test doesn't work for me - it seems that no json file is created

kevinxufs replied:
hmm that's odd. What did you get as your result from printing?
I got something like `[]` for `output_filecsv` and `['catalogue_results/202000827-163438.json']` for the `output_filejson`.

radka-j commented on catalogue/config.py:
```diff
@@ -0,0 +1,103 @@
+
+
+    with open('catalogue_config.yaml', 'w', newline='') as yaml_file:
+        writer = yaml.dump(cata_dict, yaml_file)
```
Why is this saved to a variable? I think line 102 could just be `yaml.dump(...)`

kevinxufs replied:
Thank you, I've changed this now.

Commit:
simplified some code in config

radka-j commented:
@kevinxufs ignore my last 2 comments, it was a mistake on my side. All the tests pass :)

kevinxufs commented:
Ah great thanks @radka-j for taking a look :)

Commit:
changed config doc message

edaub reviewed (approved):
Nice work, looks good and all the tests pass.

kevinxufs commented:
Thanks @edaub , does that mean it is okay to merge?

edaub commented:
Yes -- feel free to merge yourself, or I can do it if you would prefer! I think squashing this would be best, as you don't need the full history of your commits.
